text/0195-kubernetes-resource-detection.md

```diff
@@ -0,0 +1,113 @@
+
+### User Journey: Kubernetes Config Management
+
+Kubernetes templating tools, such as [Helm templates](https://helm.sh/docs/chart_template_guide/), allow users to easily re-use snippets of yaml in many places in their configuration.  A common set of environment variables would users to define the OpenTelemetry env var config once, and use it across all of their OpenTelemetry deployments.
+
+**Note:** This pattern unfortunately isn't easy to use with kubernetes' built-in patching tools, [kustomize](https://kustomize.io/), because it doesn't [currently](https://github.com/kubernetes-sigs/kustomize/issues/1493) support patching all containers across deployments with a single patch.
+
+## Internal details
+
+Add an experimental k8s.md specification under specification/resource, which specifies how kubernetes resource detectors should detect the kubernetes semantic conventions:
+
+SDK's should add a Kubernetes environment variable-based resource detector.  The Kubernetes resource detector and the collector's `resourcedetectionprocessor` should support the following environment variables, and map them to the following semantic conventions:
+
+| Environment Variable | Semantic Convention |
+| ---- | ---- |
+| K8S_POD_NAME | k8s.pod.name |
+| K8S_POD_UID | k8s.pod.uid |
+| K8S_NAMESPACE_NAME | k8s.namespace.name |
+| K8S_NODE_NAME | k8s.node.name |
+
+Detection of the above semantic conventions using environment variables should not be done inside vendor-specific (e.g. GKE-specific) resource detectors.
+
+## Prior art and alternatives
+
+### Alternative: Detection required by SDKs
+
+Similar to `OTEL_RESOURCE_ATTRIBUTES`, we could require the proposed environment variables to always be detected.  While this would further improve the ease of use, [Resource semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/README.md) (including Kubernetes) are still experimental, and the [SDK environment variable specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/sdk-environment-variables.md#general-sdk-configuration) are stable. Additionally, Kubernetes is [explicitly called out](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/sdk.md#detecting-resource-information-from-the-environment) as a custom resource detector.
+
+### Alternative: Using HOSTNAME for k8s.pod.name
+
+Many kubernetes detectors currently use `HOSTNAME` environment variable, which defaults to the Pod name. However, the `HOSTNAME` can be [modified in a few ways](https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pod-s-hostname-and-subdomain-fields) in the pod spec. Kubernetes resource detectors may fall back to detecting the pod name using `HOSTNAME` if `K8S_POD_NAME` is not available, but this may cause user confusion in some cases.
+
+### Alternative: Using Dependent Environment Variables
```

Comment:
Would this approach conflict with "Kubernetes is [explicitly called out](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/sdk.md#detecting-resource-information-from-the-environment) as a custom resource detector" above?

I wonder if having infrastructure set `OTEL_RESOURCE_ATTRIBUTES` as a means of resource detection is generally a good pattern. I know `OTEL_RESOURCE_ATTRIBUTES` is safe to self-concatenate, but it would be easy to accidentally forget to do so and override a previously set value. @mat-rumian's suggestion would improve this, but then the detection becomes less automatic.

Reply:
If we adopted this alternative, it would probably change that phrasing in the spec, which I suspect intends to use kubernetes as an example of a platform, rather than prescribe how k8s detection should be done.

I agree using the single `OTEL_RESOURCE_ATTRIBUTES` is easy to mess up, and hard to do consistently.  I'll call that out.